### DISCUSSION TITLE
How to use enlightenment Desktop with Xrdp in FreeBSD?

### DISCUSSION THREAD

**Discussion by Al89B5vm** [neutrinolabs/xrdp#2358] [Q&A]
setting Enlightenment DE in /usr/local/etc/xrdp/startwm.sh with FreeBSD in a VM ,give to me access problem i tried: exec enlightenment-session ,exec startenlightenment, and exec start-enlightenment-desktop,  following example in this script file: in ~/.xsession i typed correctly: dbus-launch --exit-with-session enlightenment_start and Desktop Environment  work properly,  not Remote Desktop connection throw neutrinoLabs display after a log in with my credential.Host OS Windows 11 home ,Hypervisor Virtualbox.which is the correct configuration?

**Comment by matt335672** [comment]
Hi @Al89B5vm 

It's hard for me to understand what you are asking here, exactly.

Can you provide a bit more detail?
- Which FreeBSD version?
- How are you installing xrdp?
- How are you installing enlightenment?
- What does your `~/.xsession` look like?

Thanks.

**Reply by Al89B5vm** [reply]
Hi @matt335672

thanks to  replay,

**FreeBSD 13.1-RELEASE-p2 amd64** ,packages xrdp xorgxrdp,

 installed with :**sudo pkg ins -y xrdp xorgxrdp** ,

 Enlightenment and xfce4,with gnome like support desktop environments,

  **xrdp_enable="YES"** and **xrdp_sesman_enable="YES"** in **/etc/rc.conf**

 in **/usr/local/etc/xrdp/xrdp.ini** ,I typed **port=tcp://:3389** instead of **port=3389** ,

and I left **use_vsock=false**,

 **/usr/local/etc/xrdp/sesman.ini** didn't  touch, 

**/usr/local/etc/xrdp/startwm.sh** first i typed **sudo cp /usr/local/etc/xrdp/startwm.sh ~/** 

then **sudo nano ~/startwm.sh**: **exec startxfce4** to try if configuration  working fine..ifconfig to know correct ip..then i opened remote desktop connection in Windows 11

 and also logged in with display-manager "gdm"inside freebsd VM ,

 they both working  fine, when i typed **"VMip:3389"** appeared  certificate request and i accepted, 

when neutrinoLabs bluesky screen appeared I typed credentials and Xfce4 desktop was smoothly and better that virtualdesktop inside Vbox

..but with enlightenment i put exec **enlightenment_start** in **~/startwm.sh**; 

and **dbus-launch --exit-with-session enlightenment _start** ,in **~/.xinitrc and ~/.xsession**

 after I typed credential remained bluesky background and nothing happened and after awhile disconnect its session,

when I just logged in with gdm , Enlightenment work fine....this is all.

 Thanks in advance if you can help me.

Sorry ,then i read a issue about enlightenment and wrong configuration in github yet reported,

 and i followed it to see if it could have solved, but looks is not.

Ask me if you need some others infos

**Comment by matt335672** [comment]
I couldn't get gdm working at all, but I started enlightenment from the console by putting the following in `~/.xinitrc` and using `startx`:-

```bash
#!/bin/sh

exec /usr/local/bin/enlightenment_start
```

Once I'd gotten enlightenment working with `startx`, I replaced `/usr/local/etc/xrdp/startwm.sh` with:-

```bash
#!/bin/sh

# Create a log file
d=${DISPLAY#*:} ;# Remove up to and including ':'
d=${d%.*} ;# Remove trailing .<screen>
exec >~/.xrdp-session.$d.log 2>&1

# Display some debug information
echo ==== Environment ===
env

# Call the user's window manager
. ~/.xinitrc
```

On my system I can connect to enlightenment over RDP doing that.

I also get the following log files:-
-  `~/.xrdp-session.10.log`
- `~/.e-log.log`

The first of these is created by my new `/usr/local/etc/xrdp/startwm.sh`, and the second is the normal enlightenment log.

Does that makes sense?
Can you try those files on your system and tell us what happens?

**Reply by Al89B5vm** [reply]
### Thanks @matt335672 for the solution

I left **~/.xinitrc** like before:
**export E_COMP_ENGINE=sw
exec dbus-launch --exit-with-session enlightenment_start**
then: in /usr/local/etc/xrdp/startwm.sh I remove # where is write:

**# This script is an example. Edit this to suit your needs.**
**If ${HOME}/startwm.sh exists, xrdp-sesman will execute it instead of this.**
and add **#** where was **exec enlightenment_start**

then: **sudo nano ~/startwm.sh** I typed:

> #!/bin/sh
> 
> **# Create a log file**
> **d=${DISPLAY#*:} ;# Remove up to and including ':'
> d=${d%.*} ;# Remove trailing .<screen>
> exec >~/.xrdp-session.$d.log 2>&1**
> 
> **# Display some debug information**
> **echo ==== Environment ===
> env**
> 
> **# Call the user's window manager**
> **. ~/.xinitrc**

i also changed xorgxrdp and xrdp with dev-packages xrdp-dev xorgxrdp-dev

and now I can look this github discussion in firefox inside Remote desktop

Thanks all works fine!!!
and both logs files are present, of course 

> LOGNAME=al89b5vm
LANG=en_US.UTF-8
XRDP_PULSE_SINK_SOCKET=xrdp_chansrv_audio_out_socket_10
UID=1001
XRDP_SESSION=1
PATH=/sbin:/bin:/usr/bin:/usr/local/bin
DISPLAY=:10.0
PWD=/usr/home/al89b5vm
XRDP_SOCKET_PATH=/var/run/xrdp
USER=al89b5vm
HOME=/home/al89b5vm
XRDP_PULSE_SOURCE_SOCKET=xrdp_chansrv_audio_in_socket_10
SHELL=/bin/sh
PULSE_SCRIPT=/usr/local/etc/xrdp/pulse/default.pa

**Comment by Al89B5vm** [comment]
well i was too impulsive, I tested your script above to launch Enlightenment session throw xrdp in a VM, inside VMware player ,freebsd 13.1-p2,quarterly branch, working fine!But when i tested now in Virtualbox, latest branch ,same release,look it don't work, only neutrinolabs bluesky screen appeared and nothing more, can log in with Display-manager, or throw console with startx, also with embedded remote desktop option inside VirtualBox as localhost with 127.0.0.1:3389, but not over rdp with xrdp .I am little confused 😕 Anyway thanks .I done again all the steps.
 **sudo rm /usr/local/etc/xrdp/xrdp.ini** , 
**sudo rm /usr/local/etc/xrdp/sesman.ini** ,
 **sudo rm /usr/local/etc/xrdp/startwm.sh**
**sudo rm ~/.xinitrc** and **~/.xsession**
**sudo pkg remove xorgxrdp  xrdp** and **sudo pkg autoremove** .
reboot system
**sudo pkg ins -y xorgxrdp xrdp**
opened **/usr/local/etc/xrdp/xrdp.ini** typed **port=tcp://:3389 instead of  port=3389**

opened sudo nano /usr/local/etc/xrdp/startwm.sh 

and removed **#** beside **If ${HOME}/startwm.sh exists, xrdp-sesman will execute it instead of this.**
then **sudo touch ~/startwm.sh** and **sudo chmod +x ~/startwm.sh**
copied your script above
then **sudo touch ~/.xinitrc** and **sudo nano ~/.xinitrc**
**exec /usr/local/bin/enlightenment_start** , same in ~/.xsession
**sudo chmod +x ~/.xinitrc**
**sudo service xrdp restart**
 I tried to log in but after i typed my credentials just a bluesky screen for awhile then report an error
![Connessione Desktop remoto 08_09_2022 14_47_51](https://user-images.githubusercontent.com/107185875/189127161-1d0085a1-25df-45a0-ab93-c1c50de02703.png)

**Reply by matt335672** [reply]
The session isn't starting, or is finishing unexpectedly.

Have a look in the two log files and see what's in there if anything.

**Reply by Al89B5vm** [reply]
**sudo Xorg :12 -auth Xauthority -config xrdp/xorg.conf -noreset -nolisten tcp -logfile .xorgxrdp.%s.log**

> cat .xorgxrdp.10.log | less

[  1427.132]
X.Org X Server 1.21.1.4
X Protocol Version 11, Revision 0
[  1427.132] Current Operating System: FreeBSD FreeBSD 13.1-RELEASE-p2 FreeBSD 13.1-RELEASE-p2 GENERIC amd64
[  1427.132]
[  1427.132] Current version of pixman: 0.40.0
[  1427.132]    Before reporting problems, check http://wiki.x.org
        to make sure that you have the latest version.
[  1427.132] Markers: (--) probed, (**) from config file, (==) default setting,
        (++) from command line, (!!) notice, (II) informational,
        (WW) warning, (EE) error, (NI) not implemented, (??) unknown.
[  1427.132] (++) Log file: ".xorgxrdp.10.log", Time: Thu Sep  8 22:05:38 2022
[  1427.132] (++) Using config file: "/etc/X11/xrdp/xorg.conf"
[  1427.132] (==) Using system config directory "/usr/local/share/X11/xorg.conf.d"
[  1427.132] (**) Option "defaultserverlayout" "X11 Server"
[  1427.132] (**) ServerLayout "X11 Server"
[  1427.132] (**) |-->Screen "Screen (xrdpdev)" (0)
[  1427.132] (**) |   |-->Monitor "Monitor"
[  1427.132] (**) |   |-->Device "Video Card (xrdpdev)"
[  1427.132] (**) |-->Input Device "xrdpMouse"
[  1427.132] (**) |-->Input Device "xrdpKeyboard"
[  1427.132] (**) Option "DontVTSwitch" "on"
[  1427.132] (**) Option "AutoAddDevices" "off"
[  1427.132] (**) Not automatically adding devices
[  1427.132] (==) Automatically enabling devices
[  1427.132] (==) Not automatically adding GPU devices
[  1427.132] (==) Automatically binding GPU devices
[  1427.132] (==) Max clients allowed: 256, resource mask: 0x1fffff
[  1427.132] (==) FontPath set to:
       /usr/local/share/fonts/misc/,
        /usr/local/share/fonts/TTF/,
        /usr/local/share/fonts/OTF/,
        /usr/local/share/fonts/Type1/,
        /usr/local/share/fonts/100dpi/,
        /usr/local/share/fonts/75dpi/,
        catalogue:/usr/local/etc/X11/fontpath.d
[  1427.132] (==) ModulePath set to "/usr/local/lib/xorg/modules"
[  1427.132] (II) Module ABI versions:
[  1427.132]    X.Org ANSI C Emulation: 0.4
[  1427.132]    X.Org Video Driver: 25.2
[  1427.132]    X.Org XInput driver : 24.4
[  1427.132]    X.Org Server Extension : 10.0
[  1427.132] (--) PCI:*(0@0:2:0) 15ad:0405:15ad:0405 rev 0, Mem @ 0xe0000000/134217728, 0xe8400000/2097152, I/O @ 0x0000c020/16, BIOS @ 0x????????/65536
[  1427.132] (II) "glx" will be loaded. This was enabled by default and also specified in the config file.
[  1427.132] (II) LoadModule: "dbe"
[  1427.132] (II) Module "dbe" already built-in
[  1427.132] (II) LoadModule: "ddc"
[  1427.132] (II) Module "ddc" already built-in
[  1427.132] (II) LoadModule: "extmod"
[  1427.132] (II) Module "extmod" already built-in
[  1427.132] (II) LoadModule: "glx"
[  1427.132] (II) Loading /usr/local/lib/xorg/modules/extensions/libglx.so
[  1427.132] (II) Module glx: vendor="X.Org Foundation"
[  1427.132]    compiled for 1.21.1.4, module version = 1.0.0
[  1427.132]    ABI class: X.Org Server Extension, version 10.0
[  1427.132] (II) LoadModule: "int10"
[  1427.132] (II) Loading /usr/local/lib/xorg/modules/libint10.so
[  1427.132] (II) Module int10: vendor="X.Org Foundation"
[  1427.132]    compiled for 1.21.1.4, module version = 1.0.0
[  1427.132]    ABI class: X.Org Video Driver, version 25.2
[  1427.132] (II) LoadModule: "record"
[  1427.132] (II) Module "record" already built-in
[  1427.132] (II) LoadModule: "vbe"
[  1427.132] (II) Loading /usr/local/lib/xorg/modules/libint10.so
[  1427.132] (II) Module int10: vendor="X.Org Foundation"
[  1427.132]    compiled for 1.21.1.4, module version = 1.0.0
[  1427.132]    ABI class: X.Org Video Driver, version 25.2
[  1427.132] (II) LoadModule: "xorgxrdp"
[  1427.132] (II) Loading /usr/local/lib/xorg/modules/libxorgxrdp.so
[  1427.132] (II) Module XORGXRDP: vendor="X.Org Foundation"
[  1427.132]    compiled for 1.21.1.4, module version = 0.2.18
[  1427.132]    ABI class: X.Org Video Driver, version 25.2
[  1427.132] xorgxrdpSetup:
[  1427.132] (II) LoadModule: "fb"
[  1427.132] (II) Module "fb" already built-in
[  1427.132] (II) LoadModule: "xrdpdev"
[  1427.132] (II) Loading /usr/local/lib/xorg/modules/drivers/xrdpdev_drv.so
[  1427.132] (II) Module XRDPDEV: vendor="X.Org Foundation"
[  1427.132]    compiled for 1.21.1.4, module version = 0.2.18
[  1427.132]    ABI class: X.Org Video Driver, version 25.2
[  1427.132] xrdpdevSetup:
[  1427.132] (II) LoadModule: "xrdpmouse"
[  1427.132] (II) Loading /usr/local/lib/xorg/modules/input/xrdpmouse_drv.so
[  1427.132] (II) Module XRDPMOUSE: vendor="X.Org Foundation"
[  1427.132]    compiled for 1.21.1.4, module version = 0.2.18
[  1427.132]    Module class: X.Org XInput Driver
[  1427.132]    ABI class: X.Org XInput driver, version 24.4
[  1427.142] rdpmousePlug:
[  1427.142] (II) LoadModule: "xrdpkeyb"
[  1427.142] (II) Loading /usr/local/lib/xorg/modules/input/xrdpkeyb_drv.so
[  1427.142] (II) Module XRDPKEYB: vendor="X.Org Foundation"
[  1427.142]    compiled for 1.21.1.4, module version = 0.2.18
[  1427.142]    Module class: X.Org XInput Driver
[  1427.142]    ABI class: X.Org XInput driver, version 24.4
[  1427.142] rdpkeybPlug:
[  1427.142] rdpIdentify:
[  1427.142] (II) XRDPDEV: driver for xrdp: XRDPDEV
[  1427.142] rdpDriverFunc: op 10
[  1427.142] (WW) Falling back to old probe method for XRDPDEV
[  1427.142] rdpProbe:
[  1427.142] (II) Loading sub module "fb"
[  1427.142] (II) LoadModule: "fb"
[  1427.142] (II) Module "fb" already built-in
[  1427.142] (II) XRDPDEV(0): using default device
[  1427.142] (WW) VGA arbiter: cannot open kernel arbiter, no multi-card support
[  1427.142] rdpPreInit:
[  1427.142] (**) XRDPDEV(0): Depth 24, (--) framebuffer bpp 32
[  1427.142] (==) XRDPDEV(0): RGB weight 888
[  1427.142] (==) XRDPDEV(0): Using gamma correction (1.0, 1.0, 1.0)
[  1427.142] (==) XRDPDEV(0): Default visual is TrueColor
[  1427.142] (==) XRDPDEV(0): DPI set to (96, 96)
[  1427.142] (II) XRDPDEV(0):   mode "640x480" ok
[  1427.142] (II) XRDPDEV(0):   mode "800x600" ok
[  1427.142] (II) XRDPDEV(0): Virtual size is 800x600 (pitch 800)
[  1427.142] (**) XRDPDEV(0):  Default mode "800x600": 36.0 MHz (scaled from 0.0 MHz), 35.2 kHz, 56.2 Hz
[  1427.142] (II) XRDPDEV(0): Modeline "800x600"x0.0   36.00  800 824 896 1024  600 601 603 625 +hsync +vsync (35.2 kHz d)
[  1427.142] rdpScreenInit: virtualX 800 virtualY 600 rgbBits 8 depth 24
[  1427.142] rdpScreenInit: pfbMemory bytes 1920000
[  1427.142] rdpScreenInit: pfbMemory 0x80183e800
[  1427.142] rdpSimdInit: assigning yuv functions
[  1427.142] rdpSimdInit: cpuid ax 1 cx 0 return ax 0x000806ec bx 0x01020800 cx 0x5eda2203 dx 0x178bfbff
[  1427.142] rdpSimdInit: sse2 amd64 yuv functions assigned
[  1427.142] (==) XRDPDEV(0): Backing store enabled
[  1427.142] rdpClientConInit: disconnect idle session after [0] sec
[  1427.142] rdpClientConInit: kill disconnected [0] timeout [0] sec
[  1427.142] rdpXvInit: depth 24
[  1427.142] rdpScreenInit: out
[  1427.142] (II) Initializing extension Generic Event Extension
[  1427.142] (II) Initializing extension SHAPE
[  1427.142] (II) Initializing extension MIT-SHM
[  1427.142] (II) Initializing extension XInputExtension
[  1427.142] (II) Initializing extension XTEST
[  1427.142] (II) Initializing extension BIG-REQUESTS
[  1427.142] (II) Initializing extension SYNC
[  1427.142] (II) Initializing extension XKEYBOARD
[  1427.142] (II) Initializing extension XC-MISC
[  1427.142] (II) Initializing extension SECURITY
[  1427.142] (II) Initializing extension XFIXES
[  1427.142] (II) Initializing extension RENDER
[  1427.142] (II) Initializing extension RANDR
[  1427.142] (II) Initializing extension COMPOSITE
[  1427.142] (II) Initializing extension DAMAGE
[  1427.142] (II) Initializing extension MIT-SCREEN-SAVER
[  1427.142] (II) Initializing extension DOUBLE-BUFFER
[  1427.142] (II) Initializing extension RECORD
[  1427.142] (II) Initializing extension DPMS
[  1427.142] (II) Initializing extension Present
[  1427.142] (II) Initializing extension DRI3
[  1427.142] (II) Initializing extension X-Resource
[  1427.142] (II) Initializing extension XVideo
[  1427.142] (II) Initializing extension XVideo-MotionCompensation
[  1427.142] (II) Initializing extension GLX
[  1427.142] (II) AIGLX: Screen 0 is not DRI2 capable
[  1427.172] (II) IGLX: Loaded and initialized swrast
[  1427.172] (II) GLX: Initialized DRISWRAST GL provider for screen 0
[  1427.172] (II) Initializing extension XFree86-VidModeExtension
[  1427.172] (II) Initializing extension XFree86-DGA
[  1427.172] (II) Initializing extension XFree86-DRI
[  1427.172] (II) Initializing extension DRI2
[  1427.172] rdpCreateScreenResources:
[  1427.223] (II) Using input driver 'XRDPMOUSE' for 'xrdpMouse'
[  1427.223] (**) Option "CorePointer"
[  1427.223] (**) xrdpMouse: always reports core events
[  1427.223] rdpmousePreInit: drv 0x801019300 info 0x8097c4f40, flags 0x0
[  1427.223] (II) XINPUT: Adding extended input device "xrdpMouse" (type: Mouse, id 6)
[  1427.223] rdpmouseControl: what 0
[  1427.223] rdpmouseDeviceInit:
[  1427.223] rdpmouseCtrl:
[  1427.223] rdpRegisterInputCallback: type 1 proc 0x800df0070
[  1427.223] (**) xrdpMouse: (accel) keeping acceleration scheme 1
[  1427.223] (**) xrdpMouse: (accel) acceleration profile 0
[  1427.223] (**) xrdpMouse: (accel) acceleration factor: 2.000
[  1427.223] (**) xrdpMouse: (accel) acceleration threshold: 4
[  1427.223] rdpmouseControl: what 1
[  1427.223] rdpmouseDeviceOn:
[  1427.223] (II) Using input driver 'XRDPKEYB' for 'xrdpKeyboard'
[  1427.223] (**) Option "CoreKeyboard"
[  1427.223] (**) xrdpKeyboard: always reports core events
[  1427.223] rdpkeybPreInit: drv 0x801019380 info 0x8097c5080, flags 0x0
[  1427.223] (II) XINPUT: Adding extended input device "xrdpKeyboard" (type: Keyboard, id 7)
[  1427.223] rdpkeybControl: what 0
[  1427.223] rdpkeybDeviceInit:
[  1427.242] rdpkeybChangeKeyboardControl:
[  1427.242] rdpkeybChangeKeyboardControl: autoRepeat on
[  1427.242] rdpRegisterInputCallback: type 0 proc 0x800df5ba0
[  1427.242] rdpkeybControl: what 1
[  1427.242] rdpkeybDeviceOn:
[  1427.242] (II) config/udev: Adding input device System mouse (/dev/input/event0)
[  1427.242] (II) AutoAddDevices is off - not adding device.
[  1427.252] (II) config/udev: Adding input device System keyboard multiplexer (/dev/input/event1)
[  1427.252] (II) AutoAddDevices is off - not adding device.
[  1427.252] (II) config/udev: Adding input device AT keyboard (/dev/input/event2)
[  1427.252] (II) AutoAddDevices is off - not adding device.
[  1427.252] (II) config/udev: Adding input device IntelliMouse Explorer (/dev/input/event3)
[  1427.252] (II) AutoAddDevices is off - not adding device.
[  1427.252] (II) config/udev: Adding input device (unnamed) (/dev/vboxguest)
[  1427.252] (II) AutoAddDevices is off - not adding device.
[  1427.252] rdpDeferredRandR:
[  1427.252] rdpResizeSession: width 1024 height 768
[  1427.252]   calling RRScreenSizeSet
[  1427.252] rdpRRScreenSetSize: width 1024 height 768 mmWidth 271 mmHeight 203
[  1427.252] rdpRRGetInfo:
[  1427.252]   screen resized to 1024x768
[  1427.252]   RRScreenSizeSet ok 1
[  1427.349] rdpInDeferredRepeatCallback:
[  1427.349] rdpkeybChangeKeyboardControl:
[  1427.349] rdpkeybChangeKeyboardControl: autoRepeat off
[  1515.373] rdpmouseControl: what 2
[  1515.373] rdpmouseDeviceOff:
[  1515.373] rdpkeybControl: what 2
[  1515.373] rdpkeybDeviceOff:
[  1515.373] rdpkeybControl: what 3
[  1515.373] rdpkeybUnInit: drv 0x801019380 info 0x8097c5080, flags 0x0
[  1515.373] rdpUnregisterInputCallback: proc 0x800df5ba0
[  1515.373] rdpmouseControl: what 3
[  1515.373] rdpmouseUnInit: drv 0x801019300 info 0x8097c4f40, flags 0x0
[  1515.373] rdpUnregisterInputCallback: proc 0x800df0070
[  1515.382] rdpCloseScreen:
[  1515.382] xorgxrdpDownDown:
[  1515.382] xorgxrdpDownDown: 1
[  1515.382] rdpClientConDeinit:
[  1515.382] rdpClientConDeinit: deleting file /tmp/.xrdp/xrdp_display_10
[  1515.382] rdpClientConDeinit: deleting file /tmp/.xrdp/xrdp_disconnect_display_10
[  1515.382] (II) Server terminated successfully (0). Closing log file.

**Maybe this is useful**
others logs:first is ever the same, and other .e-log.log is long and didn't have interesting infos

**Reply by Al89B5vm** [reply]
to launch correctly instead of use Xorg authentication , i must enabled remote desktop inside VirtualBox **port:5910 authentication method :external timeout:5000 advanced option:enable multiple connection ,then log in with ip:3389** 
**xvnc + my credential**
tested with xfce4 and  Enlightenment 
![Connessione Desktop remoto 08_09_2022 23_13_20](https://user-images.githubusercontent.com/107185875/189227539-6167ef52-c4b0-466e-bd86-2169bbeac296.png)

**Reply by matt335672** [reply]
@Al89B5vm - I'm a bit confused. Do you mean it's working OK now?

**Reply by Al89B5vm** [reply]
just i mean I need to use this workaround with this branch , when it didn't necessary before with quarterly branch,yet was a problem before reported to bsd bugzilla ,about virtual services,virtualbox-guest-additions,maybe also xrdp,have problem with Xorg,and need to report like other.thanks to your help, Good luck **@matt335672**
